### ISSUE TITLE
prefab struct predicate issue 

### ISSUE THREAD

**Issue by cderici** [pycket/pycket#275]
Here's the interaction by @samth
```
> (require syntax/parse/private/rep-data)
> rhs?
#<procedure:rhs?>
> (rhs? (rhs #f #F #F #f #F #f #f))
#t
> (rhs? #s(rhs #f #F #F #f #F #f #f))
#f 
```

It seems to be true for any prefab struct on the repl:

```
Welcome to Pycket v7.1.0.7
> (define-struct a (b c) #:prefab)
> (a? (a 1 2))
[1543965725.348079] compiling linklet : module (linklet ((.top-level-bind! .top-level-require!) 
(.mpi-vector .syntax-literals) (.namespace .phase .self .inspector .bulk-binding-registry 
.set-transformer!)) (make-a a?) (a? (make-a 1 2)))
#t
> (a? #s(a 1 2))
[1543965732.152892] compiling linklet : module (linklet ((.top-level-bind! .top-level-require!)
 (.mpi-vector .syntax-literals) (.namespace .phase .self .inspector .bulk-binding-registry 
.set-transformer!)) (a?) (a? (quote #s(a 1 2))))
#f
```

Although, here's a module that doesn't exhibit the above behavior that we see on the repl:

```
(module pref racket/base  
(define-struct a (b c) #:prefab)  
(printf "(a? (a 1 2)) : ~a \n(a? #s(a 1 2)) : ~a\n"  (a? (a 1 2)) (a? #s(a 1 2)))) 
```

**Comment by samth** [comment]
What if you try to read that struct from a string with read and open-input-string?

**Comment by samth** [comment]
That reproduces the bug:

```
(module pref '#%kernel
   (define-values
     (struct:a make-a a? a-b a-c)
     (let-values (((struct: make- ? -ref -set!)
                   (let-values ()
                     (let-values ()
                       (#%app
                        make-struct-type
                        'a
                        '#f
                        '2
                        '0
                        '#f
                        null
                        'prefab
                        '#f
                        '(0 1)
                        '#f
                        'a)))))
       (#%app
        values
        struct:
        make-
        ?
        (#%app make-struct-field-accessor -ref '0 'b)
        (#%app make-struct-field-accessor -ref '1 'c))))
   (define-values (v) (#%app read (#%app open-input-string '"#s(a 1 2)")))
   (#%app printf '"~s ~s\n" v (#%app a? v)))
```

**Comment by samth** [comment]
The problem is that the `W_StructType` embedded in the predicate `a?` and the one created by `make-prefab-struct` are not `eq?`. There's interning of prefab struct types based on the `W_PrefabKey` value in the `W_StructType.unbound_prefab_types` table, but it doesn't seem to work here.

**Comment by cderici** [comment]
Fixed by 7934f0689c4ee1a818db84153b806f71ce47d42e
